README.md
# TimeZoneLocate

This utility is the Swift port of the [APTimeZones](https://github.com/Alterplay/APTimeZones) library with the support for bundled frameworks. If your project is Objective-C, we recommend using theirs.

* For Swift 5 version, use version `~> 0.5.0`.
* For Swift 4 version, use version `~> 0.4.0`.
* For Swift 3 version, use version `~> 0.3.0`.
* For Swift 2 version, use version `~> 0.1.0`.

### Usage:

```swift

    //San Diego, CA, USA
    let location = CLLocation(latitude: 32.88, longitude: -117.15)

    //TimeZone from extension (fast, synchronous, approximate)
    var timeZone = location.timeZone
    print(timeZone)

    //or calling the class method
    timeZone = TimeZoneLocate.timeZone(location: location)
    print(timeZone)

    //if you have the country code, you can speed things up.
    timeZone = TimeZoneLocate.timeZone(location: location, countryCode: "US")

    // A MORE accurate time zone if you are connected to network. (iOS 9.0 or later)
    location.timeZone { (tz) -> (Void) in
      guard let tz = tz else { return }
      // got a more accureate timezone from the network
      print("Network TimeZone: \(tz.identifier)")
    }

    // Same thing
    TimeZoneLocate.geocodeTimeZone(location: location) { (tz) -> (Void) in
      guard let tz = tz else { return }
      print("Network TimeZone: \(tz.identifier) (class method)")
    }

```

## Installation

To install it, simply add the following line to your Podfile:

```ruby
pod "TimeZoneLocate"
```

## Authors
 * Original Objective-C Version: Sergii Kryvoblotskyi, Alterplay
 * Swift Contributor: Anthony Persaud, persaud@modernistik.com (https://www.modernistik.com)

## License

TimeZoneLocate is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
`TimeZoneLocate` adheres to [Semantic Versioning](http://semver.org/).

---
## 0.5.0
* Project and spec updates for Swift 5 and Cocoapods 1.7.1

## 0.4.3
* Adds PR #8 to include WatchOS.

## 0.4.2
* Warning fixes to static properties.

## 0.4.1
* Support for Swift 4.2

## 0.4.0
* Support for Swift 4.0
* Adds extension method ` timeZone(completion:TimeZoneLocateResult)` that uses
reverse geocoding to fetch a more accurate time zone when connected to the network.

## 0.3.1
* Fixes #5: issue relating to empty country code.

## 0.3.0
* Changes to API method names to comply better with Swift 3.0.
* Fixing string warning. [Jawwad Ahmad](https://github.com/jawwad)
* Pryamid of doom code cleanup. [Mamnun Bhuiyan](https://github.com/mamnun)

## 0.2.0
* Converted syntax to support Swift 3.0 and XCode 8.2 and later. To keep using pre Swift 3.0, use:
   `pod 'TimeZoneLocate', '= 0.1.0'`

## 0.1.0
* Initial port.


Example/TimeZoneLocate/AppDelegate.swift
//
//  AppDelegate.swift
//  TimeZoneLocate
//
//  Created by Anthony Persaud on 01/17/2016.
//  Copyright (c) 2016 Anthony Persaud. All rights reserved.
//

import UIKit
import CoreLocation
import TimeZoneLocate

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        let location = CLLocation(latitude: 32.88, longitude: -117.15)
        //helper method
        let theTimeZone = location.timeZone
        print("TimeZone: \(theTimeZone.identifier)")
        // Or the shared method
        let timeZone = TimeZoneLocate.timeZoneWithLocation(location)
        print(timeZone)
        if let timeZone = TimeZoneLocate.timeZone(location: location, countryCode: nil) {
            print(timeZone)
        }
        
        let geocoder = CLGeocoder()
        geocoder.geocodeAddressString("San Diego, CA") { (placemarks, error) -> Void in
            if error == nil, let placemarks = placemarks, let placemark = placemarks.first {
                
                let locationTimeZone = placemark.location?.timeZone
                print(locationTimeZone!)
                //of if country code is known, then
                if let location = placemark.location, let countryCode = placemark.addressDictionary?["CountryCode"] as? String {
                    let fastTimeZone = TimeZoneLocate.timeZone(location: location, countryCode: countryCode)
                    print(fastTimeZone!)
                }
                
                //Helper method

            }
        }
        
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Example/TimeZoneLocate.playground/Contents.swift
//: Playground - noun: a place where people can play

import UIKit
import MapKit
import TimeZoneLocate
import PlaygroundSupport

PlaygroundPage.current.needsIndefiniteExecution = true
var str = "Hello, TimeZone!"

let location = CLLocation(latitude: 32.88, longitude: -117.15)
//helper method
let theTimeZone = location.timeZone
print("TimeZone: \(theTimeZone.identifier)")
// Or the shared method
let timeZone = TimeZoneLocate.timeZoneWithLocation(location)
print(timeZone)

location.timeZone { (tz) -> (Void) in
    guard let tz = tz else { return }
    print("Network TimeZone: \(tz.identifier)")
    // got a more accureate timezone from the network
}

TimeZoneLocate.geocodeTimeZone(location: location) { (tz) -> (Void) in

}
let geocoder = CLGeocoder()
geocoder.geocodeAddressString("San Diego, CA") { (placemarks, error) -> Void in
    if error == nil, let placemarks = placemarks,
                     let placemark = placemarks.first,
                     let locationTimeZone = placemark.location?.timeZone {
        
        print(locationTimeZone)
        //of if country code is known, then
        if let location = placemark.location, let countryCode = placemark.isoCountryCode,
            let fastTimeZone = TimeZoneLocate.timeZone(location: location, countryCode: countryCode) {
            print(fastTimeZone)
        }
        
    }
}


Pod/Classes/TimeZoneLocate.swift
//
//  TimeZoneLocate.swift
//
//  Created by Sergii Kryvoblotskyi on 10/17/13.
//  Copyright (c) 2013 Alterplay. All rights reserved.
//

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

/*
Original Source https://github.com/Alterplay/APTimeZones
by Sergii Kryvoblotskyi on 10/17/13.
Copyright (c) 2013 Alterplay. All rights reserved.

Project Name was changed to prevent pod conflict.
Swift Conversion by Modernistik LLC.

This version featured:
* Support for Swift 3.0 syntax
* Support for Framework linking

*/
import Foundation
import CoreLocation

public typealias TimeZoneLocateResult = (_ timeZone:TimeZone?) -> (Void)

extension CLLocation {
    public var timeZone: TimeZone {
        return TimeZoneLocate.timeZoneWithLocation(self)
    }
    
    /// Fetch a more accurate time zone using reverse geocoding.
    @available(iOS 9.0, *)
    public func timeZone(completion:@escaping TimeZoneLocateResult) {
        TimeZoneLocate.geocodeTimeZone(location: self, completion: completion)
    }
}

open class TimeZoneLocate : NSObject {
    
    // MARK: Singleton
    open static let sharedInstance = TimeZoneLocate()
    open static let timeZonesDB = TimeZoneLocate.importDataBaseFromFile("timezones.json")
    
    /// Fetch a more accurate time zone using reverse geocoding. If a TimeZone is found, it is returned, otherwise nil.
    @available(iOS 9.0, *)
    open class func geocodeTimeZone(location:CLLocation, completion:@escaping TimeZoneLocateResult) {
        CLGeocoder().reverseGeocodeLocation(location) { (placemarks, error) in
            guard error == nil, let tz = placemarks?.last?.timeZone else {
                return completion(nil)
            }
            completion(tz)
        }
    }
    
    /// Get timezone with lat/lon and country code
    open class func timeZoneWithLocation(_ location:CLLocation) -> TimeZone {
        guard let closestZoneInfo = closestZoneInfo(location:location, source: TimeZoneLocate.timeZonesDB),
              let timeZone = timeZoneWithDictionary(closestZoneInfo)
            else { return TimeZone.current } //We've found nothing. Let's use system.
        return timeZone
    }
    
    /*
    Get timezone with lat/lon and country code.
    Extremely speeds up and more carefull result.
    */
    open class func timeZone(location:CLLocation, countryCode:String? = nil) -> TimeZone? {
        //Need a country code
        guard let countryCode = countryCode,
            //Filter
            let filteredZones = filteredTimeZones(countryCode:countryCode),
            //Get closest zone info
            let closestZoneInfo = closestZoneInfo(location:location, source:filteredZones),
            //get timzone
            let timeZone = timeZoneWithDictionary(closestZoneInfo)
        else { return TimeZoneLocate.timeZoneWithLocation(location)}
        
        return timeZone
    }
    
    /*
    Import from DB
    */
    open class func importDataBaseFromFile(_ fileName:String) -> [[AnyHashable: Any]] {
        let currentBundle = Bundle(for: TimeZoneLocate.self)
        
        let filePath = currentBundle.path(forResource: fileName, ofType: nil, inDirectory: "TimeZoneLocate.bundle") ??  currentBundle.path(forResource: fileName, ofType: nil)
        do {
            if let filePath = filePath,
                let jsonData = try? Data(contentsOf: URL(fileURLWithPath: filePath)),
                let timeZones = try JSONSerialization.jsonObject(with: jsonData, options: .allowFragments) as? [[AnyHashable: Any]]
            {
                return timeZones
            }
        } catch let error as NSError {
            NSLog("Invalid timezoneDB format %@", error.localizedDescription)
        }
        
        //let filePath = currentBundle.resourcePath
        assertionFailure("Error loading or parse timeZoneDB file: \(filePath ?? "filePath is nil")")
        return [[AnyHashable: Any]]()
    }
    
    /*
    Calculates the closest distance from source
    */
    open class func closestZoneInfo(location: CLLocation, source:[[AnyHashable: Any]]?) -> [AnyHashable: Any]? {
        
        var closestDistance: CLLocationDistance = Double.infinity
        var closestZoneInfo: [AnyHashable: Any]?
        
        guard let source = source else { return nil }
            
        for locationInfo in source {
            guard let latitude = locationInfo["latitude"] as? Double,
                  let longitude = locationInfo["longitude"] as? Double else { continue }
            
            let distance = location.distance( from: CLLocation(latitude: latitude, longitude: longitude) )
            if  distance < closestDistance {
                closestDistance = distance
                closestZoneInfo = locationInfo
            }
        }
        
        return closestZoneInfo
    }
    
    /*
    Filtering the whole DB with the country code
    */
    open class func filteredTimeZones(countryCode: String) -> [[AnyHashable: Any]]? {
        let predicate = NSPredicate(format: "country_code LIKE %@", countryCode)
        return (TimeZoneLocate.timeZonesDB as NSArray).filtered(using: predicate) as? [[AnyHashable: Any]]
    }
    
    /*
    Timezone from dict
    */
    open class func timeZoneWithDictionary(_ zoneInfo: [AnyHashable: Any]?) -> TimeZone? {
        guard let zoneName = zoneInfo?["zone"] as? String else { return nil }
        
        return TimeZone(identifier: zoneName)
    }
    
}

